### DISCUSSION TITLE
Saving using the kiota-abstractions-go/serialization Package

### DISCUSSION THREAD

**Discussion by dadams39** [microsoftgraph/msgraph-sdk-go#141] [Q&A] [answered]
# How can SDK be used to write retrieved items?  
Related to: https://github.com/microsoftgraph/msgraph-sdk-go/issues/140
In the code example: https://docs.microsoft.com/en-gb/graph/sdks/paging?tabs=Go

The snippet highlights how to receive each Message from the Mailbox. Messages are defined [here][]

The [SerializationWriter Interface](https://github.com/microsoft/kiota-abstractions-go/blob/main/serialization/serialization_writer.go) is defined within Kiota abstractions, but it is unclear which how to save a particular message to an app.
Code is adapted from provided code above:
```
target := "My Message"
// Iterate over all pages (e.g. Messages)
iterateErr := pageIterator.Iterate(func(pageItem interface{}) bool {
    message := pageItem.(models.Messageable)
    fmt.Printf("%s\n", *message.GetSubject())
    if target == *message.GetSubject() {
        entries := message.GetFieldDeserializers(). // MAP Key: <string> Type: func(serialization.ParseNode) error
        for key, value := in range entries{
            // NEED CODE FOR HERE
            val() // NOT VALID     
    // Return true to continue the iteration
    return true
})
```
How do you get the value out of the value?
Golang produces the following error:
```
not enough arguments in call to val
        have ()
        want (serialization.ParseNode)
```
If the code for val is changed to ` val( message )`, the error becomes:

```
cannot use message (variable of type models.Messageable) as type serialization.ParseNode in argument to val:
        models.Messageable does not implement serialization.ParseNode (missing GetBoolValue method)
       
```
This conflict seems to exist in the [Kiota Library](https://github.com/microsoft/kiota-abstractions-go/). I'm unsure if changing the code in the `models` directory will satisfy the complaint or if it requires a change to the serialization framework in Kiota *or* if it could be the case that I'm not seeing something the actual solution.

**Comment by dadams39** [comment]
Updates to the discussion have been made in:
https://github.com/microsoftgraph/msgraph-sdk-go/issues/140

For the TL;dr, the  Kiota abstractions _deep-dive_ pages are located here:
https://microsoft.github.io/kiota/extending/serialization.html
https://microsoft.github.io/kiota/extending/models.html
While I am able to get the raw bytes of a message, it is still unclear how to take a Page Item and copy it to a local variable. Reparsing the bytes into an object does seem like going backward.

**Reply by baywet** [reply]
Are you trying to assign the subject of the first message to the target variable? Or am I missing something?

**Reply by dadams39** [reply]
This is the flow process. 
- Model looks scans the entire mailbox for subjects matching a criterion (spam)
- If the message subject is a match, all of the fields of the message are to be saved to a file which I documented in Issue #140. 
- After all the messages have been saved, the messages in question need to be but into `[]Messages`. (At this point, they are bytes... See Issue #148). 

It is unclear how to perform the last step. That is my question.

**Reply by baywet** [reply]
Thanks for clarifying the intent, I'll provide an answer we can mark.

**Comment by baywet** [comment, marked as answer]
```golang
target := "My Message"
// Iterate over all pages (e.g. Messages)
iterateErr := pageIterator.Iterate(func(pageItem interface{}) bool {
    message := pageItem.(models.Messageable)
    fmt.Printf("%s\n", *message.GetSubject())
    if target == *message.GetSubject() {
        writer, err := requestAdapter.GetSerializationWriterFactory().GetSerializationWriter("application/json")
        err := writer.WriteObjectValue("", message)
        serializedValue, err := writer.GetSerializedContent()
        // do something with the serialized value like persist it or send to another process etc...
        parseNode, err := absser.DefaultParseNodeFactoryInstance.GetRootParseNode("application/json", serializedValue)
        deserializedValue, err := parseNode.GetObjectValue(models.CreateMessageFromDiscriminatorValue)
       // do something with the deserialized value
    // Return true to continue the iteration
    return true
})
```

absser is an import of github.com/microsoft/kiota-abstractions-go/serialization
models of msgraph-go-sdk/models

> Note: you should look into odata query parameters (filter) to see whether the filtering can be done service side for you as opposed to client side.

**Reply by dadams39** [reply]
@baywet,

I am receiving this failure during this line of code
```go
parseNode, err := absser.DefaultParseNodeFactoryInstance.GetRootParseNode("application/json", serializedValue)
err >>> invalid character 'C' after object key:value pair
```
I changed a few items on the serializedValue to keep my data private, but here is the serialized value:
```
{"id":"AAMkAGQ1NzViZTdhLTEwMTMtNGJjNi05YWI2LTg4NWRlZDA2Y2UxOABGAAAAAAAPvVwUramXT7jlSGpVU8_7BwB8wYc0thTTTYl3RpEYIUq_AAAAAAEMAAB8wYc0thTTTYl3RpEYIUq_AAACXyeJAAA=","@odata.etag":"W/"CQAAABYAAAB8
wYc0thTTTYl3RpEYIUq+AAACXq4h"""body":{"content":"Thanks for taking the time!                                                                                                                   
                                                                                                                                                                                               
From: Some Person <some.person@9q3vij.onmicrosoft.com>                                                                                                                                         
Date: Friday, April 8, 2022 at 2:06 PM                                                                                                                                                         
To: That Person <another@github.ai>                                                                                                                                                              
Subject: Start                                                                                                                                                                                 
Send one                                                                                                                                                                                       
","contentType":"TEXT"},"sender":{"emailAddress":{"address":"another@github.ai","name":"That Person"}},"subject":"Re: Start"}
``` 
Would this mean that Issue #148 is still valid 🤷 ?

**Reply by baywet** [reply]
I logged an issue in the json serialization lib that's a bit more specific. https://github.com/microsoft/kiota-serialization-json-go/issues/11
Would you be willing to submit a pull request to address it?

**Reply by dadams39** [reply]
I see there is a PR already in the repo. I will look at those specific changes. In regards to the `microsoft/kiota-serialization-json-go` repository, do you have a preferred naming convention for test files, @baywet ?

**Reply by baywet** [reply]
For the naming convention just name_test.go
